File: android/src/main/java/com/fusion5/capacitorforegroundlocationservice/GeoUtils.java
package com.fusion5.capacitorforegroundlocationservice;

public class GeoUtils {

    // Calculates distance in meters between two lat/lng points using Haversine formula
    public static double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final double R = 6371000; // Radius of the Earth in meters

        double dLat = deg2rad(lat2 - lat1);
        double dLon = deg2rad(lon2 - lon1);

        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
                + Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2))
                * Math.sin(dLon / 2) * Math.sin(dLon / 2);

        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        double distance = R * c;

        return distance;
    }

    private static double deg2rad(double deg) {
        return deg * (Math.PI / 180);
    }
}

File: ios/Sources/CapacitorForegroundLocationServicePlugin/GeoUtils.swift
class GeoUtils {
    
    /// Calculates distance in meters between two coordinates using Haversine formula
    static func calculateDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let R = 6_371_000.0 // Earth's radius in meters

        let dLat = deg2rad(lat2 - lat1)
        let dLon = deg2rad(lon2 - lon1)

        let a = sin(dLat / 2) * sin(dLat / 2) +
                cos(deg2rad(lat1)) * cos(deg2rad(lat2)) *
                sin(dLon / 2) * sin(dLon / 2)

        let c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return R * c
    }

    private static func deg2rad(_ deg: Double) -> Double {
        return deg * .pi / 180
    }
}